feat: scrollable table

surajsharma opened the pull request:
#478 
ETA 16.5 hours

surajsharma commented:
@predrag-us please see if it matches design 😰

surajsharma commented:
@predrag-us i forgot to make the adjustments you have suggested in #478, will share the link presently

surajsharma commented:
hi @georgeciubotaru the CI pipeline here is toggling off the `PUB_USE_PRODUCTION_API` in `.env`, as a result, we cannot visit `/c/learn/crypto-and-web3-events` on staging, pls help 

@predrag-us once we have sorted the above problem, the table should be available for your testing purposes, i'm not quite there yet in terms of fixing everything you have mentioned, but apart from few width issues, it is ready for your review

georgeciubotaru commented:
@surajsharma The thread is available under

`/c/elysio/crypto-events`

surajsharma commented:
@predrag-us pls check out if this seems ok https://holdex-venture-studio-n6w4lyuuz-holdex-accelerator.vercel.app/c/elysio/crypto-events

predrag-us commented:
@surajsharma 

Please recheck the requirements in the issue once again, one by one. The implementation doesn't satisfy all the requirements. If you can't notice the problems, I'll help. 

Also, the fade should only be visible if there's something more to scroll to on either left or right, not all the time as implemented.

surajsharma commented:
thanks @predrag-us just wanted you to have the link for future ref. i will add the fade toggle. please do me a favor and uncheck any of the previous issues you are able to find here https://github.com/holdex/holdex-venture-studio/issues/478 

as for everything else not in the list, please leave a list of your observations here. imho only table width should be the remaining issue.

surajsharma commented:
> @surajsharma Done.
> 
> Also, notice this: ![CleanShot 2023-08-21 at 11 03 01@2x](https://user-images.githubusercontent.com/18451941/261974490-2ca79e8a-ace5-48c8-8ece-35d66f93bb6a.png)

unable to interpret the screenshot, looks like the gradient squeezed into a tiny space

surajsharma commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -2,18 +2,53 @@ mixin table-cell
+                +else
+                  +if('colIndex === 1')
+                    th(
+                      class!="w-[200px] bg-l2 text-left py-1.5 px-3 text-footnote font-bold border-solid border-l border-t border-l4 first:rounded-tl-xl last:rounded-tr-xl border-r {isLeftEnd && w>550 ? '' : 'bg-gradient-to-r to-transparent from-s1'}"
```
tailwind selector defined `mobile` isn't working as expected, hence the hard-coded number here, pls help @georgeciubotaru

georgeciubotaru replied:
@surajsharma What is the issue here?

Look at my screenshot
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/bbc5aebe-ae7c-4d12-b71d-c180fda91907)

surajsharma replied:
@georgeciubotaru not able to reporoduce this, neither in the browser console nor in the code, need to check why. will come back to this after #493

surajsharma commented:
@predrag-us please take a look at the table in https://holdex-venture-studio-n5jv1pknc-holdex-accelerator.vercel.app/c/elysio/crypto-events and kindly share your thoughts. is the width ok on large screens (604/710 px)? and also is the scroll working as expected. there may be a couple code-related issues here but kindly ensure there are no surprises.

predrag-us reviewed (changes requested):
@surajsharma Please have a look at the unchecked items in #478. They're not solved. 

Also, there's a problem with the fade:
- The left one overlaps the column divider and is visible at all times, which shouldn't be the case. 
- The right one is not visible when scrolling but is stuck at the end of the table.  

Remember the idea behind the fade ⏤ indicate that there's something more on the left/right.

surajsharma commented:
> @surajsharma Please have a look at the unchecked items in #478. They're not solved.
> The width on large screens is not okay as it should be 710px.

this is baffling! looking into it 

> 
> Also, there's a problem with the fade:
> 
> * The left one overlaps the column divider and is visible at all times

pls, tell me are you seeing this fade even at >550px width? thats what i understand by "at all times" here. i can see if there is a way to remove the column border. 

> * The right one is not visible when scrolling but is stuck at the end of the table.

let me see if there is a way to achieve this behaviour inside an html table, i assumed the fade was to be shown only on scroll end. with what you're suggesting, does it mean that the right scroll disappears upon scrolling all the way to the right?

predrag-us commented:
@surajsharma 

> pls, tell me are you seeing this fade even at >550px width? thats what i understand by "at all times" here. i can see if there is a way to remove the column border.

I'm talking about this:
![CleanShot 2023-08-24 at 11 01 20@2x](https://github.com/holdex/holdex-venture-studio/assets/18451941/ccaab3c0-9199-485c-a123-fabfd8439f84)
As you can notice, the border between the first and second column disappeared. 

> with what you're suggesting, does it mean that the right scroll disappears upon scrolling all the way to the right?

That's right. When you're on either one of the ends, left or right, the fade shouldn't be visible since there's nothing more to scroll to.

surajsharma commented:
@predrag-us 

> I'm talking about this: ![CleanShot 2023-08-24 at 11 01 20@2x](https://user-images.githubusercontent.com/18451941/262921159-ccaab3c0-9199-485c-a123-fabfd8439f84.png) As you can notice, the border between the first and second column disappeared.

you're right, i will add back the border

> > with what you're suggesting, does it mean that the right scroll disappears upon scrolling all the way to the right?
> 
> That's right. When you're on either one of the ends, left or right, the fade shouldn't be visible since there's nothing more to scroll to.

understood, will fix

georgeciubotaru commented:
@surajsharma Leave a log here. Let's merge this ASAP

surajsharma commented:
@georgeciubotaru there are some final touches that need work here, 1-2 hours more should do the work on my end, will complete after #493

surajsharma commented:
@predrag-us 

- have changed the logic of scroll-shades to "start of scroll" instead of end of scroll
- fixed the sticky col border issue


looks like this to me, far from perfect, but almost there, what say?

https://github.com/holdex/holdex-venture-studio/assets/5298624/c8bb10a6-3135-4d5b-8a92-b158d39c7d7b

surajsharma commented:
@predrag-us @georgeciubotaru 

fixed all issues here, tested on mobile, i feel this is ready, please have a look

georgeciubotaru commented on tailwind.config.cjs:
```diff
@@ -75,6 +75,7 @@ module.exports = {
       t2: 'rgb(var(--t2) / <alpha-value>)',
       t3: 'rgb(var(--t3) / <alpha-value>)',
       t4: 'rgb(var(--t4) / <alpha-value>)',
+      s1: 'rgb(var(--s1) / <alpha-value>)',
```
same here

surajsharma replied:
these are the colors for shades on the scrolling table used in `src/lib/components/BodyRenderer/Blocks/table.pug`

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -2,18 +2,54 @@ mixin table-cell
+  use:scrollAction
+)
+
+.relative.overflow-auto(use:scrollAction)
```
@surajsharma We have scrollable table implemented in Clearpool, just see the example and perform the same logic here

surajsharma replied:
@georgeciubotaru i will look at the clearpool implementation and adapt this one

surajsharma replied:
removed. resolving this comment.

georgeciubotaru reviewed (changes requested):
see my comments

predrag-us reviewed (changes requested):
@surajsharma 

1. The border issue between the first two columns isn't solved. 
2. There's a new problem now as the edge of the table disappears when not scrolled to the end (screenshot).
3. You don't need the fade at the very end on the right since there's nothing to scroll to.
4. You need the fade before the end on the right since there's something to scroll to. 
5. Please have a look at the unchecked items in #478 as they're still not solved. Seems like you're having issues understanding them. Am I right?

![CleanShot 2023-08-28 at 10 30 33@2x](https://github.com/holdex/holdex-venture-studio/assets/18451941/39540a4b-7c0b-455b-a69f-40f281c37c80)

surajsharma commented:
> @surajsharma
> 
> 1. The border issue between the first two columns isn't solved.

the sticky col border is now visible on scrolling
<img width="174" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/28514a22-e700-4dc1-972f-f26fc5778c49">



> 2. There's a new problem now as the edge of the table disappears when not scrolled to the end (screenshot).

what is the expected behaviour here? the design does not show in mobile view

> 3. You don't need the fade at the very end on the right since there's nothing to scroll to.

there is no fade once you scroll all the way in either direction

> 4. You need the fade before the end on the right since there's something to scroll to.

there is fade in the direction in which srolling is possible


> 5. Please have a look at the unchecked items in [Problem: The table isn't implemented by the design #478](https://github.com/holdex/holdex-venture-studio/issues/478) as they're still not solved. Seems like you're having issues understanding them. Am I right?


i think there is some technical problem here i am working on it, but help me understand: the table is `710px` or `610px` wide in the non-scrollable large screen view? which size is applied when?

predrag-us commented:
@surajsharma 

> what is the expected behaviour here? the design does not show in mobile view

The borders of the table should always be visible. Only the content of the table should be scrollable. 

> i think there is some technical problem here i am working on it, but help me understand: the table is 710px or 610px wide in the non-scrollable large screen view? which size is applied when?

710px, as in the design.

surajsharma commented:
@predrag-us 
> The borders of the table should always be visible. Only the content of the table should be scrollable.

got it

> 710px, as in the design.

but the parent container is only 604pixels wide, so do we want to increase the whole parent width? or somehow have the wider (`710`px) table place itself absolutely over the narrow parent? hope i am being clear here..

surajsharma commented on src/lib/styles/variables.scss:
```diff
@@ -23,6 +23,7 @@ $light-colors: (
   t2: toSpacedRgb(#2f323d),
   t3: toSpacedRgb(#5d657a),
   t4: toSpacedRgb(#ffffff),
+  s1: toSpacedRgb(#e0e4ec),
```
these are the colors for shades on the scrolling table used in `src/lib/components/BodyRenderer/Blocks/table.pug`

georgeciubotaru replied:
you already have this color

surajsharma replied:
removed. resolving this comment.

surajsharma commented:
@predrag-us didn't make sense (and was getting complex) to display a table wider than the parent, so i fixed by increasing parent width, but is that what we wanted? please take a look, all other issues are addressed as well.

predrag-us commented:
@surajsharma 

Please notice the chart on the landing page. It's 710px and centered. That's how the table should behave as well.

surajsharma commented:
@predrag-us 

> Please notice the chart on the landing page. It's 710px and centered. That's how the table should behave as well.

i had guessed as much but still wanted to confirm for clarity, thanks!

surajsharma commented:
@predrag-us @georgeciubotaru have addressed all pending issues here, pls take a look

predrag-us reviewed (changes requested):
@surajsharma 

The issues mentioned [here](https://github.com/holdex/holdex-venture-studio/pull/496#pullrequestreview-1597794895), [here](https://github.com/holdex/holdex-venture-studio/pull/496#issuecomment-1685936759), and [here](https://github.com/holdex/holdex-venture-studio/pull/496#pullrequestreview-1593060206) are still unresolved. 

I've unchecked the items in #478 that aren't solved.

surajsharma commented:
@predrag-us getting the width to behave without breaking scroll is proving to be a challenge, a design compromise here could be keeping it at 610 for now.. not saying i cannot achieve this push of the envelope here as well, but this pr is already 4-in-1, which is abnormal.

surajsharma commented:
@predrag-us 

tracking #478 for requirements, my thoughts below

> Check the cell content alignment (should be top-left instead of centered)

it is top-left

> The table width on large screens is inaccurate (should be 710px instead of 604px). I even noticed a table with 655.22px which shouldn't be the case. Our dimensions are 710px and 604px | [first one](https://holdex.io/c/learn/crypto-and-web3-events)

it is 710/604px till the nearest breakpoint

> We should have a `max-width` for the columns. My recommendation is 200px. So, when we have a longer text, don't try to squeeze it in (for example, available space) 100px but allow it to span the full 200px

all cols now support `max-width` of `200px`

> The tables should be scrollable on the x-axis while the first column should be fixed. Notice the fades in the design, after the first column and after the last column, indicating there's more on the left and right. So, don't try to squeeze everything in 710px, but allow scrolling on the x-axis while the column `max-width` is 200px and the first column has a fixed position

this has been achieved / should be per your satisfaction now and working as expected unless i am wrong, kindly take a look at this video and tell me which parts are not working for you https://www.youtube.com/watch?v=6xD76XZGloI

---

not tracking the stale concerns:

> The issues mentioned [here](https://github.com/holdex/holdex-venture-studio/pull/496#pullrequestreview-1597794895), [here](https://github.com/holdex/holdex-venture-studio/pull/496#issuecomment-1685936759), and [here](https://github.com/holdex/holdex-venture-studio/pull/496#pullrequestreview-1593060206) are still unresolved.
> 


pls share further issues here

predrag-us reviewed (changes requested):
@surajsharma 

> Predrag: We should have a max-width for the columns. My recommendation is 200px. So, when we have a longer text, don't try to squeeze it in (for example, available space) 100px but allow it to span the full 200px
> Suraj: all cols now support max-width of 200px

No, they don't. 

For example, the column here has a width of 70px while it clearly has enough content to span longer than 70px.

<img width="308" alt="CleanShot 2023-08-31 at 10 12 33@2x" src="https://github.com/holdex/holdex-venture-studio/assets/18451941/74dfd2a2-8c83-4940-91f8-b1c7ff66c39f">

--- 

- When I'm in the middle of the table, both sides should have fades, since there's more content on both sides. This is not the case now. 
- I can't see the right border of the table when I'm not on the end of the table. 

--- 

- Please notice how the top and bottom borders are affected by the fade. That shouldn't be the case. 

![CleanShot 2023-08-31 at 10 09 09@2x](https://github.com/holdex/holdex-venture-studio/assets/18451941/2521f911-3bf0-429c-9bc9-9b38aea81ec5)

--- 

- Please notice how Links don't have any padding on top and bottom. 

<img width="126" alt="CleanShot 2023-08-31 at 10 11 05@2x" src="https://github.com/holdex/holdex-venture-studio/assets/18451941/7b52b2ce-2a36-4d26-aca0-e04c8fdb75f7">

--- 

This issue is still valid, as the border is visible only at the very start and very end.

<img src="https://user-images.githubusercontent.com/18451941/263645742-39540a4b-7c0b-455b-a69f-40f281c37c80.png">

surajsharma commented:
> No, they don't.
> 
> For example, the column here has a width of 70px while it clearly has enough content to span longer than 70px.
> 
> <img alt="CleanShot 2023-08-31 at 10 12 33@2x" width="308" src="https://user-images.githubusercontent.com/18451941/264580470-74dfd2a2-8c83-4940-91f8-b1c7ff66c39f.png">

> * When I'm in the middle of the table, both sides should have fades, since there's more content on both sides. This is not the case now.

please explain this further, because from your description it sounds as if you want me to force each cell to be 200px wide instead of letting the table decide for itself..

> * I can't see the right border of the table when I'm not on the end of the table.


fixed

> * Please notice how the top and bottom borders are affected by the fade. That shouldn't be the case.
> ![CleanShot 2023-08-31 at 10 09 09@2x](https://user-images.githubusercontent.com/18451941/264579586-2521f911-3bf0-429c-9bc9-9b38aea81ec5.png)


please explain this, you want the sticky border to be rounded on the right as well? not clear from description what is needed here...

> * Please notice how Links don't have any padding on top and bottom.

fixed 


> <img alt="CleanShot 2023-08-31 at 10 11 05@2x" width="126" src="https://user-images.githubusercontent.com/18451941/264580056-7b52b2ce-2a36-4d26-aca0-e04c8fdb75f7.png">
> This issue is still valid, as the border is visible only at the very start and very end.
> 
> ![](https://user-images.githubusercontent.com/18451941/263645742-39540a4b-7c0b-455b-a69f-40f281c37c80.png)

fixed: border is visible at all times for me

surajsharma commented:
@predrag-us 

> > No, they don't.
> > For example, the column here has a width of 70px while it clearly has enough content to span longer than 70px.
> > <img alt="CleanShot 2023-08-31 at 10 12 33@2x" width="308" src="https://user-images.githubusercontent.com/18451941/264580470-74dfd2a2-8c83-4940-91f8-b1c7ff66c39f.png">


i do not think this is possible without using another library. for the rest of the points this is good to go 

cc @georgeciubotaru

surajsharma commented:
as i said [before](https://github.com/holdex/holdex-venture-studio/pull/496#issuecomment-1704551151) giving a table cell a dynamic width based on content length needs at the very least, some external lib or extra code to caculate table width which is beyond scope. however, i have added a min and max width to give a similar effect. please see if ok. @georgeciubotaru @predrag-us

georgeciubotaru commented:
@surajsharma Sync your branch with the main branch.

1. First update your local `main-branch`
2. merge if into your branch

surajsharma commented:
@georgeciubotaru done and resolved conflicts

georgeciubotaru commented:
@surajsharma 

1. git checkout `main`
2. git pull
3. git checkout `478-table-redesign`
4. git merge --no-ff main
5. resolve conflicts
6. push the change

surajsharma commented:
@georgeciubotaru done

georgeciubotaru commented on src/routes/(pages)/layout.scss:
```diff
@@ -3,3 +3,26 @@
+  @apply xl:-left-[53px] lg:-left-[53px] md:-left-[53px] tb:-left-[53px];
+}
+
+td {
+  vertical-align: top;
+  text-align: left;
+  @apply p-3 w-full;
+}
```
this one can me moved in HTML

surajsharma replied:
@georgeciubotaru no, if i use it html it does not work

surajsharma replied:
@georgeciubotaru i made it all tailwind in a single `apply`

georgeciubotaru replied:
@surajsharma you can move to HTML why to keep in CSS that

surajsharma replied:
@georgeciubotaru alright, done

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -2,18 +2,52 @@ mixin table-cell
+
++table-scroll-shades
+.relative.overflow-auto(use:scrollAction class="rounded-xl" id="table-parent")
+  table(class="border-separate border-spacing-0" id="table-scroll")
+    +each('cells as row, rowIndex')
+      thead
+        tr
+          +each('row as cellContents, colIndex')
+            +if('rowIndex === 0')
+              +if('colIndex === 0')
+                th(
+                  class="col-fixed-width text-left left-0 sticky bg-l2 px-3 text-footnote font-bold border-solid border-l border-t border-l4 first:rounded-tl-xl"
+                  scope="col"
+                )
+                  +table-cell
+                +else
+                  th(
+                    class="col-mw bg-l2 text-left py-1.5 px-3 text-footnote font-bold border-solid border-l last:border-r border-t border-l4 first:rounded-tl-xl last:rounded-tr-xl"
+                    scope="col"
+                  )
+                    +table-cell
+      tbody
+        tr
+          +each('row as cellContents, colIndex')
+            +if('rowIndex != 0')
+              +if('colIndex === 0')
+                td(
+                  class="z-20 bg-l1 left-0 sticky px-3 text-t2 text-footnote border-solid border-l border-t border-l4 last:border-r {rowIndex === cells.length - 1 ? 'first:rounded-bl-xl last:rounded-br-xl border-b': ''}"
+                )
+                  +table-cell
+                +else
+                  td(
+                    class="col-mw bg-l1 py-1.5 px-3 text-footnote border-solid border-l border-t border-l4 last:border-r {rowIndex === cells.length - 1 ? 'first:rounded-bl-xl last:rounded-br-xl border-b': ''} "
+                  )
+                    +table-cell
```
can you optimise the code and make it DRY, use mixins

surajsharma replied:
@georgeciubotaru ok how about now?

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,59 @@
+mixin table-row(rowIndex, colIndex, cellContents)
+  +if('rowIndex != 0')
+    +if('colIndex === 0')
+      td(
+        class="z-20 bg-l1 left-0 sticky px-3 text-t2 text-footnote border-solid border-l border-t border-l4 last:border-r {rowIndex === cells.length - 1 ? 'first:rounded-bl-xl last:rounded-br-xl border-b': ''}"
+      )
```
the idea was to optimize this element, because it has approx the same classes as the line 10

surajsharma replied:
@georgeciubotaru although they have almost the same classes it would not make sense to further compartmentalise here because the combination of classes is unique. i can make separate mixins for each but i see no point, unless you insist...

georgeciubotaru replied:
you can have something like

mixin tableDataCell()&attributes(attributes)
	td(class="<add_here_common_clasess>")
	
	
usage
+tableDataCell()(class!="<difference_classes>"(

surajsharma replied:
this is fixed

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,59 @@
+mixin table-head-row(rowIndex, colIndex, cellContents)
+  +if('rowIndex === 0')
+    +if('colIndex === 0')
+      th(
+        class="col-fixed-width text-left left-0 sticky bg-l2 px-3 text-footnote font-bold border-solid border-l border-t border-l4 first:rounded-tl-xl"
+        scope="col"
+      )
```
same here

surajsharma replied:
@georgeciubotaru pls see above comment

surajsharma replied:
fixed

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,30 +1,30 @@
-        class="col-fixed-width text-left left-0 sticky bg-l2 px-3 text-footnote font-bold border-solid border-l border-t border-l4 first:rounded-tl-xl"
-        scope="col"
-      )
+      +table-header-cell.div(class!="col-fixed-width left-0 sticky")
```
Why do you need this `div`?

+table-header-cell()(class="!col-fixed-width !left-0 !sticky")

surajsharma replied:
it autoformats to remove the first ()

surajsharma commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,59 @@
+      tbody
+        tr
+          +each('row as cellContents, colIndex')
+            +table-row(rowIndex, colIndex, cellContents)
```
@georgeciubotaru fixed

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,61 @@
+        +table-data-cell
+
+mixin table-head-row(rowIndex, colIndex, cellContents)
+  +if('rowIndex === 0')
```
then you don't need this anymore I think

surajsharma replied:
@georgeciubotaru fixed

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,61 @@
+.relative.overflow-auto(use:scrollAction class="rounded-xl" id="table-parent")
+  table(class="border-separate border-spacing-0" id="table-scroll")
+    +each('cells as row, rowIndex')
+      +if('rowIndex===0')
+        thead
+          tr
+            +each('row as cellContents, colIndex')
+              +table-head-row(rowIndex, colIndex, cellContents)
+        +else
+          tbody
+            tr
+              +each('row as cellContents, colIndex')
+                +table-row(rowIndex, colIndex, cellContents)
```
@surajsharma What will be the issue if you will have the same structure as in the previous version?
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/d4cf0577-5293-4bd7-a80f-1c3d69265a4d)

georgeciubotaru replied:
Please first all issues then ask for review, please check the `Inspector`  tab to see the HTML structure

surajsharma replied:
@georgeciubotaru this https://user-images.githubusercontent.com/8217518/266013856-1322e855-6b86-46cc-a105-39d83eea2c3a.png

surajsharma replied:
@georgeciubotaru table structure is now as expected 
<img width="308" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/1b7348d1-ac1b-43cc-8873-b4778c3731e7">

georgeciubotaru replied:
I've provided the version in screeshot

surajsharma replied:
@georgeciubotaru fixed

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/table.pug:
```diff
@@ -1,19 +1,57 @@
+    +table-cell
+
+mixin table-row(rowIndex, colIndex, cellContents)
+  +if('rowIndex != 0')
+    +if('colIndex === 0')
+      +table-data-cell.div(class!="z-20 left-0 sticky text-t2")
+      +else
+        +table-data-cell
```
```suggestion
  +if('colIndex === 0')
    +table-data-cell.div(class!="z-20 left-0 sticky text-t2")
    +else
      +table-data-cell
```
